velox/substrait/VeloxToSubstraitExpr.cpp

```diff
@@ -0,0 +1,311 @@
+    substraitExpr->mutable_literal()->MergeFrom(
+        toSubstraitExpr(arena, constExpr));
+    return *substraitExpr;
+  }
+  if (auto callTypeExpr =
+          std::dynamic_pointer_cast<const CallTypedExpr>(expr)) {
+    substraitExpr->MergeFrom(toSubstraitExpr(arena, callTypeExpr, inputType));
+    return *substraitExpr;
+  }
+  if (auto fieldExpr =
+          std::dynamic_pointer_cast<const FieldAccessTypedExpr>(expr)) {
+    substraitExpr->mutable_selection()->MergeFrom(
+        toSubstraitExpr(arena, fieldExpr, inputType));
+
+    return *substraitExpr;
+  }
+  if (auto castExpr = std::dynamic_pointer_cast<const CastTypedExpr>(expr)) {
+    substraitExpr->mutable_cast()->MergeFrom(
+        toSubstraitExpr(arena, castExpr, inputType));
+
+    return *substraitExpr;
+  }
+
+  VELOX_UNSUPPORTED("Unsupport Expr '{}' in Substrait", expr->toString());
+}
+
+const ::substrait::Expression_Cast&
+VeloxToSubstraitExprConvertor::toSubstraitExpr(
+    google::protobuf::Arena& arena,
+    const std::shared_ptr<const CastTypedExpr>& castExpr,
+    const RowTypePtr& inputType) {
+  ::substrait::Expression_Cast* substraitCastExpr =
+      google::protobuf::Arena::CreateMessage<::substrait::Expression_Cast>(
+          &arena);
+  std::vector<TypedExprPtr> castExprInputs = castExpr->inputs();
+
+  substraitCastExpr->mutable_type()->MergeFrom(
+      typeConvertor_->toSubstraitType(arena, castExpr->type()));
+
+  for (auto& arg : castExprInputs) {
+    substraitCastExpr->mutable_input()->MergeFrom(
+        toSubstraitExpr(arena, arg, inputType));
+  }
+  return *substraitCastExpr;
+}
+
+const ::substrait::Expression_FieldReference&
+VeloxToSubstraitExprConvertor::toSubstraitExpr(
+    google::protobuf::Arena& arena,
+    const std::shared_ptr<const FieldAccessTypedExpr>& fieldExpr,
+    const RowTypePtr& inputType) {
+  ::substrait::Expression_FieldReference* substraitFieldExpr =
+      google::protobuf::Arena::CreateMessage<
+          ::substrait::Expression_FieldReference>(&arena);
+
+  const auto& exprType = fieldExpr->type();
+  std::string exprName = fieldExpr->name();
+
+  ::substrait::Expression_ReferenceSegment_StructField* directStruct =
+      substraitFieldExpr->mutable_direct_reference()->mutable_struct_field();
+
+  directStruct->set_field(inputType->getChildIdx(exprName));
+
+  return *substraitFieldExpr;
+}
+
+const ::substrait::Expression& VeloxToSubstraitExprConvertor::toSubstraitExpr(
+    google::protobuf::Arena& arena,
+    const std::shared_ptr<const CallTypedExpr>& callTypeExpr,
+    const RowTypePtr& inputType) {
+  ::substrait::Expression* substraitExpr =
+      google::protobuf::Arena::CreateMessage<::substrait::Expression>(&arena);
+
+  auto inputs = callTypeExpr->inputs();
+  auto& functionName = callTypeExpr->name();
+
+  // The processing is different for different function names.
+  // TODO add support for if Expr and switch Expr
+  if (functionName != exec::kIf && functionName != exec::kSwitch) {
```

Comment:
It might be cleaner to look up functionName in functionMap_ and throw "unsupported" if not found.

Reply:
This is because for substrait, `exec::kIf` need to be converted to`::substrait::Expression_IfThen` and for `exec::kSwitch`, it need to be converted to `::substrait::Expression_SwitchExpression`. And here we only have the case `::substrait::Expression_ScalarFunction`,  so that we add if statement here. Do you mean this or may you have any suggestions?

Reply:
Got it. Thank you for clarifying.